crates/vm/src/system/memory/offline.rs

```diff
@@ -25,6 +25,91 @@ struct BlockData {
     size: usize,
 }
 
+struct BlockMap {
+    /// Block ids. 0 is a special value standing for the default block.
+    id: AddressMap<usize, PAGE_SIZE>,
+    /// The place where non-default blocks are stored.
+    storage: Vec<BlockData>,
+    initial_block_size: usize,
+}
+
+impl BlockMap {
+    pub fn from_mem_config(mem_config: &MemoryConfig, initial_block_size: usize) -> Self {
+        assert!(initial_block_size.is_power_of_two());
+        Self {
+            id: AddressMap::from_mem_config(mem_config),
+            storage: vec![],
+            initial_block_size,
+        }
+    }
+
+    fn initial_block_data(pointer: u32, initial_block_size: usize) -> BlockData {
+        let aligned_pointer = (pointer / initial_block_size as u32) * initial_block_size as u32;
+        BlockData {
+            pointer: aligned_pointer,
+            size: initial_block_size,
+            timestamp: INITIAL_TIMESTAMP,
+        }
+    }
+
+    pub fn get_without_adding(&self, address: &(u32, u32)) -> BlockData {
+        let idx = self.id.get(address).unwrap_or(&0);
+        if idx == &0 {
+            Self::initial_block_data(address.1, self.initial_block_size)
+        } else {
+            self.storage[idx - 1].clone()
+        }
+    }
+
+    pub fn get(&mut self, address: &(u32, u32)) -> &BlockData {
+        let (address_space, pointer) = *address;
+        let idx = self.id.get(&(address_space, pointer)).unwrap_or(&0);
+        if idx == &0 {
+            let pointer = pointer - pointer % self.initial_block_size as u32;
```

Comment:
Is this faster than `pointer / N * N`, or is it just stylistic?

I wonder if we should do `pointer & ~(N - 1)` (assuming `N` is a power of 2). If we make things generic over const `N` I think the compiler can compile this to one instruction.